diesel/src/query_builder/update_statement/changeset.rs

```diff
@@ -68,20 +69,95 @@ where
-        out.push_identifier(T::NAME)?;
+        QueryFragment::walk_ast(&self.target, out.reborrow())?;
         out.push_sql(" = ");
-        QueryFragment::walk_ast(&self.expr, out)
+        QueryFragment::walk_ast(&self.expr, out.reborrow())
+    }
+}
+
+/// Represents the left hand side of an assignment expression for an
+/// assignment in [AsChangeset]. The vast majority of the time, this will
+/// be a [Column]. However, in certain database backends, it's possible to
+/// assign to an expression. For example, in Postgres, it's possible to
+/// "UPDATE TABLE SET array_column\[1\] = 'foo'".
+///
+/// You shouldn't have to implement this trait. Expressions can be made
+/// assignable with [AssignableExpression].
+pub trait AssignmentTarget {
+    /// Table the assignment is to
+    type Table: Table;
+    /// A wrapper around a type to assign to (this wrapper should implement
+    /// [QueryFragment]).
+    type QueryAstNode;
+
+    /// Move this in to the AST node which should implement [QueryFragment].
+    fn into_target(self) -> Self::QueryAstNode;
+}
+
+/// Represents a `Column` as an `AssignmentTarget`. The vast majority of
+/// targets in an update statement will be `Column`s.
+#[derive(Debug, Clone, Copy)]
+pub struct ColumnWrapperForUpdate<C>(pub C);
+
+impl<DB, C> QueryFragment<DB> for ColumnWrapperForUpdate<C>
+where
+    DB: Backend + DieselReserveSpecialization,
+    C: Column,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        out.push_identifier(C::NAME)
+    }
+}
+
+/// Trait to indicate that this type is valid to be the assignment
+/// of an update statement.
+pub trait AssignableExpression {
+    /// An assignable expression is effectively operating on a transformation
+    /// of a single column signified by this type.
+    type Column: Column;
+    /// The expression to assign to. It's normally required to transform
+    /// `Column`s using [crate::query_builder::UncorrelatedColumn]. This
+    /// removes the table identifier.
+    type Expression;
+
+    /// Return the expression in a form which can be used in an update
+    /// statement.
+    fn into_assignable_expression(self) -> Self::Expression;
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ExpressionWrapperForUpdate<C>(C);
+
+impl<DB, Expr> QueryFragment<DB> for ExpressionWrapperForUpdate<Expr>
+where
+    DB: Backend + DieselReserveSpecialization,
+    Expr: QueryFragment<DB> + AssignableExpression,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        QueryFragment::walk_ast(&self.0, out.reborrow())
+    }
+}
+
+impl<Expr> AssignmentTarget for Expr
+where
+    Expr: AssignableExpression,
+{
+    type Table = <<Expr as AssignableExpression>::Column as Column>::Table;
+    type QueryAstNode = ExpressionWrapperForUpdate<Expr::Expression>;
+
+    fn into_target(self) -> Self::QueryAstNode {
+        ExpressionWrapperForUpdate(self.into_assignable_expression())
```

Comment:
Is there a specific reason here why we implement this for generic expressions, instead of just for `ArrayIndex<C, Expr>`? Fixing this would likely remove the conflicting impl error with the generic column impl, which in turn would simplify the whole implementation as well.

Reply:
Yes - that's a fair point. This doesn't need 2 trait layers.